### DISCUSSION TITLE
Recursive BOM listing

### DISCUSSION THREAD

**Discussion by lucobos** [inventree/InvenTree#3681] [Q&A]
Hi,

Using InvenTree 0.8.2

We have 9 products based on the same pcb. For this pcb we have defined 10 sets of components:
-Set Common
-Sets Input12, Input24 and Input48
-Sets Output12, Output24 and Output48

and products are 12-12, 12-24, 12-48, 24-12 and so on.

So each product is formed by Inputxx, Common and Outputxx.
Each set has a number of parts and has been assigned a BOM.

Product 12-24 is made from Common, Input12 and Ouput24.

In InvenTree I have defined each set (Common, Input12 and Ouput24) as an Assembly and Component Part.
And the part 12-24 as an Assembly of Common, Input12 and Ouput24.

So far, so good.

But when I make a build for 12-24 I do not get the list of individual single parts unless I make a build of Commom, Input12 and Output24 before.
Since Common, Inputxx and Outputxx are not real parts but sets of components, I cannot "create" these parts.

In fact, what I want to avoid is listing each set individually so, when building 12-24, all parts of Common, Input12 and Output24 are listed as a single BOM though they are 3 intermediate BOMs.

I want an automatic recursive procedure. How should I manage this usecase?

Thanks,
Pedro.

**Comment by SchrodingersGat** [comment]
@lucobos recursive builds must be handled sequentially. First build the sub-components and then use those in the higher level assembly. In-line building of "recursive" builds or BOMs is not supported. Implementing this would throw up a lot of extra complications

**Comment by matmair** [comment]
@lucobos while this is an easy way to implement the parts it is not really reflecting how the system works. Parts are designed to reflect the reality of the assembly.

Creating what you ask for is possible but would require some work code-wise. Can you code in python?
We normally only work on features we can personally use so I am not sure if someone will work on this.

**Reply by matmair** [reply]
@lucobos what you could do to reduce the duplication is to use part variants for the common parts.

**Comment by lucobos** [comment]
@SchrodingersGat , @matmair 
Thanks Oliver, Matthias,
I wanted to know whether this possibility already existed. I had tried marking the "set" part to virtual too.
Sure, I can do some coding or pay someone to do it but I have not (yet) learnt anything about the code of InvenTree which seems a very complex artwork (sic). I prefer to have the system implemented in my company before trying to introduce any change in the program.

I will explore the idea of variants, though my example of 3 inputs, 3 outputs is a simple one. For some dc-dc converters we have 36 pcb variants (same board, different parts) with 5 different enclosures. I was thinking of variants for the enclosures but for the pcb I have to re-work the idea. I am the newcomer in the company and they have worked with sets for years. I can't break _all_ the workflows at the same time but one at a time.

Thanks again for your help.

**Reply by matmair** [reply]
@lucobos do you have further questions or can we close this one as answered?
